harness/determined/core/_checkpoint.py

```diff
@@ -34,6 +34,107 @@ class DownloadMode(enum.Enum):
     NoSharedDownload = "NO_SHARED_DOWNLOAD"
 
 
+def merge_metadata(all_metadata: List[Dict[str, Any]]) -> Tuple[Dict[str, Any], Dict[str, List]]:
+    """
+      Given a list of metadata, return:
+        - merged metadata
+        - a dict mapping conflicting keys to ranks that would upload them
+
+    Merging scenarios:
+    #  Metadata 1        | Metadata 2          | Conflict |  Result
+    # -------------------|---------------------|----------|-------------------------------
+    #  key1: val         | key2: val           | no       | {key1: val, key2: val}
+    #  key1: val         | key1: val           | yes      | n/a
+    #  key1: [val]       | key1: [val]         | no       | {key1: [val, val]}
+    #  key1: [val]       | key1: val           | yes      | n/a
+    #  key1: [val]       | key1: {key3: val}   | yes      | n/a
+    #  key1: {key3: val} | key1: {key4: val}   | no       | key1: {key3: val, key4: val}
+    #  key1: {key3: val} | key1: val           | yes      | n/a
+    #  key1: {key3: val} | key1: [val]         | yes      | n/a
+    """
+    merged: Dict[str, Any] = {}
+    conflicts: Dict[str, List] = {}
+    for rank, rank_metadata in enumerate(all_metadata):
+
+        for _key in rank_metadata:
+            metadata = rank_metadata[_key]
+            if _key not in merged:
+                merged[_key] = metadata
+                conflicts[_key] = [rank]
+            else:
+                if not isinstance(metadata, type(merged[_key])):
+                    # if values under the same keys have different types
+                    # report conflict and skip merging
+                    conflicts[_key].append(rank)
+                else:
+                    if isinstance(metadata, list):
+                        # merge two lists
+                        merged[_key].extend(metadata)
+                    elif isinstance(metadata, dict):
+                        # recursively merge dictionaries and collect conflicts
+                        merged[_key], nested_conflicts = merge_metadata([merged[_key], metadata])
```

Comment:
Stylistic note: the nature of the recursion seems... fishy to me.

The entrypoint to the recursive function is "we have these N different ranks each reporting an object (which might be mergeable)".  But the point where you recurse is more like "here's something that might have been merged and another entry to merge into it".  I would instead expect you to collect all such objects, and recurse only once for a given `key` within the overall function body.

The fact that it is fishy is not really a problem, per se, so long as it is correct and doesn't introduce a lot of extra work.  Is one call to `merge_metadata` with `N` entries equal to `N - 1` calls to `merge_metadata` with 2 entries?  ... That's hard for me to think about, especially because there's unbounded depth to the recursion after the asymmetry is introduced.

It's not the end of the world if the implementation of a single function is hard to think about, so long as it is correct.  But I suspect if you wrote the recursion differently, it might be a lot easier to think about.

Reply:
Fixed. I added a test with your example and one more with multiple conflicts at different levels.